File: scala/func/TreeNode.scala
package func

import scala.collection.mutable.ArrayBuffer
import scala.util.control.Breaks.{break, breakable}

class TreeNode(_value: Int = 0, _left: TreeNode = null, _right: TreeNode = null) {
  var value: Int = _value
  var left: TreeNode = _left
  var right: TreeNode = _right
}


object TreeNode {
  def from(arrayWithNulls: Any*): TreeNode = {
    if (arrayWithNulls.isEmpty) {
      return null
    }

    val array = arrayWithNulls.map(value => if (value == null) {
      None
    } else {
      Some(value.asInstanceOf[Int])
    }).toList

    val root = new TreeNode(array.head.get)

    var previousRow = ArrayBuffer[TreeNode](root)
    var i = 1

    breakable {
      while (i < array.size) {
        val currentRow = ArrayBuffer[TreeNode]()
        for (node <- previousRow) {
          if (array(i).isDefined) {
            node.left = new TreeNode(array(i).get)
            currentRow.append(node.left)
          }
          i += 1
          if (i >= array.size) {
            break
          }
          if (array(i).isDefined) {
            node.right = new TreeNode(array(i).get)
            currentRow.append(node.right)
          }
          i += 1
        }
        previousRow = currentRow
      }
    }

    root
  }
}

File: typescript/func/TreeNode.ts
class TreeNode {
  val: number;
  left: TreeNode | null;
  right: TreeNode | null;

  constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
    this.val = val === undefined ? 0 : val;
    this.left = left === undefined ? null : left;
    this.right = right === undefined ? null : right;
  }

  static fromArray(array: Array<number | null>): TreeNode | null {
    if (array.length === 0 || array[0] === null) {
      return null;
    }

    const root = new TreeNode(array[0]);
    let previousRow: TreeNode[] = [root];

    let i = 1;
    while (i < array.length) {
      const currentRow: TreeNode[] = [];
      for (const node of previousRow) {
        if (array[i] !== null) {
          node.left = new TreeNode(array[i]!);
          currentRow.push(node.left);
        }
        i += 1;
        if (i >= array.length) {
          break;
        }

        if (array[i] !== null) {
          node.right = new TreeNode(array[i]!);
          currentRow.push(node.right);
        }
        i += 1;
      }
      previousRow = currentRow;
    }

    return root;
  }
}

export default TreeNode;
